flink-test-utils-parent/flink-connector-testing-framework/src/main/java/org/apache/flink/connectors/test/common/testsuites/TestSuiteBase.java

```diff
@@ -0,0 +1,430 @@
+        LOG.debug("Writing {} records to external system", testRecordCollection.size());
+        externalContext.createSourceSplit().writeRecords(testRecordCollection);
+        return testRecordCollection;
+    }
+
+    /**
+     * Check if the given test environment is controllable (can trigger failover or network
+     * isolation).
+     *
+     * @param testEnvironment Test environment being checked
+     * @throws IllegalArgumentException if the test environment is not controllable
+     */
+    protected void checkEnvironmentIsControllable(TestEnvironment testEnvironment)
+            throws IllegalArgumentException {
+        assumeTrue(
+                ClusterControllable.class.isAssignableFrom(testEnvironment.getClass()),
+                "Provided test environment should be controllable.");
+    }
+
+    /**
+     * Check if records consumed by Flink are identical to test records written into the single
+     * split within the limit of record number.
+     *
+     * @param testRecordIterator Iterator of test records
+     * @param resultRecordIterator Iterator of result records consumed by Flink
+     * @param limit Number of records to check
+     */
+    protected void checkSingleSplitRecords(
+            Iterator<T> testRecordIterator, Iterator<T> resultRecordIterator, int limit) {
+        for (int i = 0; i < limit; i++) {
+            T testRecord = testRecordIterator.next();
+            T resultRecord = resultRecordIterator.next();
+            assertEquals(testRecord, resultRecord);
+        }
+        LOG.debug("{} records are validated", limit);
+    }
+
+    /**
+     * Check if records consumed by Flink are identical to test records written into the single
+     * split.
+     *
+     * @param testRecordIterator Iterator of test records
+     * @param resultRecordIterator Iterator of result records consumed by Flink
+     */
+    protected void checkSingleSplitRecords(
+            Iterator<T> testRecordIterator, Iterator<T> resultRecordIterator) {
+        int recordCounter = 0;
+
+        while (testRecordIterator.hasNext()) {
+            T testRecord = testRecordIterator.next();
+            T resultRecord = resultRecordIterator.next();
+            assertEquals(testRecord, resultRecord);
+            recordCounter++;
+        }
+
+        assertFalse(resultRecordIterator.hasNext());
+        LOG.debug("{} records are validated", recordCounter);
+    }
+
+    /**
+     * Check if records consumed by Flink is identical to test records written into multiple splits.
+     *
+     * <p>The order of records across different splits can be arbitrary, but should be identical
+     * within a split.
+     *
+     * @param testRecordIterators Collection of iterators for records in different splits
+     * @param resultRecordIterator Iterator of result records consumed by Flink
+     */
+    protected void checkMultipleSplitRecords(
+            Collection<Iterator<T>> testRecordIterators, Iterator<T> resultRecordIterator) {
```

Comment:
I'd probably materialize all results and get rid of iterators. It's just so much harder to read...